diff --git a/src/printf.c b/src/printf.c
index a07e741..3aa5f0b 100644
--- a/src/printf.c
+++ b/src/printf.c
@@ -1,7 +1,7 @@
 /* Formatted output functions (printf functions family).
 
-Copyright 2007-2023 Free Software Foundation, Inc.
-Contributed by the AriC and Caramba projects, INRIA.
+Copyright 2007-2025 Free Software Foundation, Inc.
+Contributed by the Pascaline and Caramba projects, INRIA.
 
 This file is part of the GNU MPFR Library.
 
@@ -16,9 +16,8 @@ or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 License for more details.
 
 You should have received a copy of the GNU Lesser General Public License
-along with the GNU MPFR Library; see the file COPYING.LESSER.  If not, see
-https://www.gnu.org/licenses/ or write to the Free Software Foundation, Inc.,
-51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA. */
+along with the GNU MPFR Library; see the file COPYING.LESSER.
+If not, see <https://www.gnu.org/licenses/>. */
 
 #ifdef HAVE_CONFIG_H
 # include "config.h"
@@ -32,25 +31,27 @@ https://www.gnu.org/licenses/ or write to the Free Software Foundation, Inc.,
 
 #include "mpfr-impl.h"
 
-#ifdef _MPFR_H_HAVE_FILE
-
 /* Each printf-like function calls mpfr_vasnprintf_aux (directly or
    via mpfr_vasprintf), which
    - returns the number of characters to be written excluding the
      terminating null character (disregarding the size argument);
    - returns -1 and sets the erange flag if this number exceeds INT_MAX
-     (in that case, also sets errno to EOVERFLOW on POSIX systems). */
+     (in that case, also sets errno to EOVERFLOW on POSIX systems).
+
+   Moreover, since the output may contain non-terminating null characters
+   (if %c is used with the value 0), the mpfr_free_str function must not be
+   used to free the allocated memory, because the size may matter with some
+   custom allocation functions. Anyway, mpfr_free_func is more efficient
+   here, as the size does not need to be recomputed. */
+
+#ifdef _MPFR_H_HAVE_FILE
 
 #define GET_STR_VA(sz, str, fmt, ap)            \
   do                                            \
     {                                           \
       sz = mpfr_vasprintf (&(str), fmt, ap);    \
       if (sz < 0)                               \
-        {                                       \
-          if (str)                              \
-            mpfr_free_str (str);                \
-          return -1;                            \
-        }                                       \
+        return -1;                              \
     } while (0)
 
 #define GET_STR(sz, str, fmt)                   \
@@ -61,11 +62,16 @@ https://www.gnu.org/licenses/ or write to the Free Software Foundation, Inc.,
       sz = mpfr_vasprintf (&(str), fmt, ap);    \
       va_end (ap);                              \
       if (sz < 0)                               \
-        {                                       \
-          if (str)                              \
-            mpfr_free_str (str);                \
-          return -1;                            \
-        }                                       \
+        return -1;                              \
+    } while (0)
+
+#define MPFR_FPRINTF_END(ret, str, fp)          \
+  do                                            \
+    {                                           \
+      int status;                               \
+      status = fwrite ((str), (ret), 1, (fp));  \
+      mpfr_free_func ((str), (ret) + 1);        \
+      return status == 1 ? (ret) : -1;          \
     } while (0)
 
 int
@@ -75,10 +81,7 @@ mpfr_printf (const char *fmt, ...)
   int ret;
 
   GET_STR (ret, str, fmt);
-  ret = printf ("%s", str);
-
-  mpfr_free_str (str);
-  return ret;
+  MPFR_FPRINTF_END (ret, str, stdout);
 }
 
 int
@@ -88,10 +91,7 @@ mpfr_vprintf (const char *fmt, va_list ap)
   int ret;
 
   GET_STR_VA (ret, str, fmt, ap);
-  ret = printf ("%s", str);
-
-  mpfr_free_str (str);
-  return ret;
+  MPFR_FPRINTF_END (ret, str, stdout);
 }
 
 
@@ -102,10 +102,7 @@ mpfr_fprintf (FILE *fp, const char *fmt, ...)
   int ret;
 
   GET_STR (ret, str, fmt);
-  ret = fprintf (fp, "%s", str);
-
-  mpfr_free_str (str);
-  return ret;
+  MPFR_FPRINTF_END (ret, str, fp);
 }
 
 int
@@ -115,14 +112,19 @@ mpfr_vfprintf (FILE *fp, const char *fmt, va_list ap)
   int ret;
 
   GET_STR_VA (ret, str, fmt, ap);
-  ret = fprintf (fp, "%s", str);
-
-  mpfr_free_str (str);
-  return ret;
+  MPFR_FPRINTF_END (ret, str, fp);
 }
 
 #endif /* _MPFR_H_HAVE_FILE */
 
+#define MPFR_SPRINTF_END(ret, buf, str)         \
+  do                                            \
+    {                                           \
+      memcpy ((buf), (str), (ret) + 1);         \
+      mpfr_free_func ((str), (ret) + 1);        \
+      return (ret);                             \
+    } while (0)
+
 int
 mpfr_sprintf (char *buf, const char *fmt, ...)
 {
@@ -130,10 +132,7 @@ mpfr_sprintf (char *buf, const char *fmt, ...)
   int ret;
 
   GET_STR (ret, str, fmt);
-  ret = sprintf (buf, "%s", str);
-
-  mpfr_free_str (str);
-  return ret;
+  MPFR_SPRINTF_END (ret, buf, str);
 }
 
 int
@@ -143,10 +142,7 @@ mpfr_vsprintf (char *buf, const char *fmt, va_list ap)
   int ret;
 
   GET_STR_VA (ret, str, fmt, ap);
-  ret = sprintf (buf, "%s", str);
-
-  mpfr_free_str (str);
-  return ret;
+  MPFR_SPRINTF_END (ret, buf, str);
 }
 
 int
diff --git a/src/vasprintf.c b/src/vasprintf.c
index b910544..ab22da6 100644
--- a/src/vasprintf.c
+++ b/src/vasprintf.c
@@ -1,8 +1,8 @@
 /* mpfr_vasnprintf_aux -- helper function for the formatted output functions
    (printf functions family).
 
-Copyright 2007-2023 Free Software Foundation, Inc.
-Contributed by the AriC and Caramba projects, INRIA.
+Copyright 2007-2025 Free Software Foundation, Inc.
+Contributed by the Pascaline and Caramba projects, INRIA.
 
 This file is part of the GNU MPFR Library.
 
@@ -17,9 +17,8 @@ or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 License for more details.
 
 You should have received a copy of the GNU Lesser General Public License
-along with the GNU MPFR Library; see the file COPYING.LESSER.  If not, see
-https://www.gnu.org/licenses/ or write to the Free Software Foundation, Inc.,
-51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA. */
+along with the GNU MPFR Library; see the file COPYING.LESSER.
+If not, see <https://www.gnu.org/licenses/>. */
 
 /* If the number of output characters is larger than INT_MAX, the
    ISO C99 / C11 standards are silent, but POSIX[*] requires the
@@ -528,10 +527,16 @@ typedef wint_t mpfr_va_wint;
       }                                         \
   } while (0)
 
-/* Process the format part which does not deal with mpfr types,
+/* Process the format part that does not deal with mpfr types,
+   from start to end (not included).
    Jump to external label 'error' if gmp_asprintf return -1.
    Note: start and end are pointers to the format string, so that
    size_t is the best type to express the difference.
+   Warning! Since the output from gmp_vasprintf may contain non-terminating
+   null characters (if %c is used with the value 0), the mpfr_free_str
+   function must not be used to free the allocated memory, because the size
+   may matter with some custom allocation functions. Anyway, mpfr_free_func
+   is more efficient here, as the size does not need to be recomputed.
    FIXME: If buf.size = 0 or size != 0, gmp_vsnprintf should be called
    instead of gmp_vasprintf, outputting data directly to the buffer
    when applicable.
@@ -548,7 +553,7 @@ typedef wint_t mpfr_va_wint;
                                                                         \
         MPFR_TMP_MARK (marker);                                         \
         fmt_copy = (char *) MPFR_TMP_ALLOC (n + 1);                     \
-        strncpy (fmt_copy, (start), n);                                 \
+        memcpy (fmt_copy, (start), n);                                  \
         fmt_copy[n] = '\0';                                             \
         length = gmp_vasprintf (&s, fmt_copy, (ap));                    \
         if (length < 0)                                                 \
@@ -557,7 +562,7 @@ typedef wint_t mpfr_va_wint;
             goto error;                                                 \
           }                                                             \
         buffer_cat ((buf_ptr), s, length);                              \
-        mpfr_free_str (s);                                              \
+        mpfr_free_func (s, length + 1);                                 \
         (flag) = 0;                                                     \
         MPFR_TMP_FREE (marker);                                         \
       }                                                                 \
@@ -645,8 +650,11 @@ buffer_widen (struct string_buffer *b, size_t len)
   MPFR_ASSERTD (*b->curr == '\0');
 }
 
-/* Concatenate the first len characters of the string s to the buffer b and
-   expand it if needed. Return non-zero if overflow. */
+/* Concatenate the first len characters of the array s to the buffer b,
+   and expand it if needed. Return non-zero if overflow.
+   Warning! The array s may contain null characters in addition to the
+   terminating one, in case %c has been used with the value 0.
+ */
 static int
 buffer_cat (struct string_buffer *b, const char *s, size_t len)
 {
@@ -657,8 +665,6 @@ buffer_cat (struct string_buffer *b, const char *s, size_t len)
   if (len == 0)
     return 0;
 
-  MPFR_ASSERTD (len <= strlen (s));
-
   if (buffer_incr_len (b, len))
     return 1;
 
@@ -669,13 +675,11 @@ buffer_cat (struct string_buffer *b, const char *s, size_t len)
       if (MPFR_UNLIKELY (b->curr + len >= b->start + b->size))
         buffer_widen (b, len);
 
-      /* strncat is similar to strncpy here, except that strncat ensures
-         that the buffer will be null-terminated. */
-      strncat (b->curr, s, len);
+      MPFR_ASSERTD (*b->curr == '\0');
+      memcpy (b->curr, s, len);
       b->curr += len;
-
       MPFR_ASSERTD (b->curr < b->start + b->size);
-      MPFR_ASSERTD (*b->curr == '\0');
+      *b->curr = '\0';
     }
 
   return 0;
@@ -2083,7 +2087,10 @@ sprnt_fp (struct string_buffer *buf, mpfr_srcptr p,
    (b) or ptr = NULL, and it implements mpfr_vsnprintf (Buf, size, fmt, ap)
    It returns the number of characters that would have been written had 'size'
    been sufficiently large, not counting the terminating null character, or -1
-   if this number is too large for the return type 'int' (overflow).
+   if this number is too large for the return type 'int' (overflow), in which
+   case, if ptr <> NULL, the memory is deallocated (otherwise, there is no way
+   to deallocate it later, since the actual size, needed by mpfr_free_func, is
+   unknown as there may be non-terminating null characters).
 */
 int
 mpfr_vasnprintf_aux (char **ptr, char *Buf, size_t size, const char *fmt,
@@ -2431,23 +2438,21 @@ mpfr_vasnprintf_aux (char **ptr, char *Buf, size_t size, const char *fmt,
   nbchar = buf.len;
   MPFR_ASSERTD (nbchar >= 0);
 
+  /* Warning! Be careful that the buffer may contain null characters
+     in addition to the terminating one, in case %c has been used with
+     the value 0. */
+
   if (ptr != NULL)  /* implement mpfr_vasprintf */
     {
-      MPFR_ASSERTD (nbchar == strlen (buf.start));
       *ptr = (char *) mpfr_reallocate_func (buf.start, buf.size, nbchar + 1);
     }
   else if (size != 0)  /* implement mpfr_vsnprintf */
     {
-      if (nbchar < size)
-        {
-          strncpy (Buf, buf.start, nbchar);
-          Buf[nbchar] = '\0';
-        }
-      else
-        {
-          strncpy (Buf, buf.start, size - 1);
-          Buf[size-1] = '\0';
-        }
+      /* The size is limited to int (see above). */
+      int len = nbchar < size ? nbchar : size - 1;
+
+      memcpy (Buf, buf.start, len);
+      Buf[len] = '\0';
       mpfr_free_func (buf.start, buf.size);
     }
 
diff --git a/tests/tfprintf.c b/tests/tfprintf.c
index 6028cb5..c0cdbf1 100644
--- a/tests/tfprintf.c
+++ b/tests/tfprintf.c
@@ -1,7 +1,7 @@
 /* tfprintf.c -- test file for mpfr_fprintf and mpfr_vfprintf
 
-Copyright 2008-2023 Free Software Foundation, Inc.
-Contributed by the AriC and Caramba projects, INRIA.
+Copyright 2008-2025 Free Software Foundation, Inc.
+Contributed by the Pascaline and Caramba projects, INRIA.
 
 This file is part of the GNU MPFR Library.
 
@@ -16,9 +16,8 @@ or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 License for more details.
 
 You should have received a copy of the GNU Lesser General Public License
-along with the GNU MPFR Library; see the file COPYING.LESSER.  If not, see
-https://www.gnu.org/licenses/ or write to the Free Software Foundation, Inc.,
-51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA. */
+along with the GNU MPFR Library; see the file COPYING.LESSER.
+If not, see <https://www.gnu.org/licenses/>. */
 
 /* FIXME: The output is not tested (thus coverage data are meaningless).
    For instance, slightly changing the code of mpfr_fprintf does not
@@ -405,6 +404,20 @@ bug_20090316 (FILE *fout)
   mpfr_clear (x);
 }
 
+/* See check_null test in tsprintf.c for details.
+   Fixed in commits
+     390e51ef8570da4e338e9806ecaf2d022210d951 (2023-12-03)
+     3babf029fe604c08ec517ca6945a5efb155f69d1 (2023-12-13)
+*/
+static void
+check_null (FILE *fout)
+{
+  int n;
+
+  check_vfprintf (fout, ".%c%c.%n", 0, 1, &n);
+  check_length (40, n, 4, d);
+}
+
 int
 main (int argc, char *argv[])
 {
@@ -442,6 +455,7 @@ main (int argc, char *argv[])
   check_random (fout, N);
 
   bug_20090316 (fout);
+  check_null (fout);
 
   fclose (fout);
   tests_end_mpfr ();
diff --git a/tests/tsprintf.c b/tests/tsprintf.c
index b85dbb9..8d13c50 100644
--- a/tests/tsprintf.c
+++ b/tests/tsprintf.c
@@ -1,8 +1,8 @@
 /* tsprintf.c -- test file for mpfr_sprintf, mpfr_vsprintf, mpfr_snprintf,
    and mpfr_vsnprintf
 
-Copyright 2007-2023 Free Software Foundation, Inc.
-Contributed by the AriC and Caramba projects, INRIA.
+Copyright 2007-2025 Free Software Foundation, Inc.
+Contributed by the Pascaline and Caramba projects, INRIA.
 
 This file is part of the GNU MPFR Library.
 
@@ -17,23 +17,8 @@ or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 License for more details.
 
 You should have received a copy of the GNU Lesser General Public License
-along with the GNU MPFR Library; see the file COPYING.LESSER.  If not, see
-https://www.gnu.org/licenses/ or write to the Free Software Foundation, Inc.,
-51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA. */
-
-/* Note: If you use a C99-compatible implementation and GMP (or MPIR)
- * has been compiled without HAVE_VSNPRINTF defined, then this test
- * may fail with an error like
- *   repl-vsnprintf.c:389: GNU MP assertion failed: len < total_width
- *
- * The reason is that __gmp_replacement_vsnprintf does not support %a/%A,
- * even though the C library supports it.
- *
- * References:
- *   https://sympa.inria.fr/sympa/arc/mpfr/2022-10/msg00001.html
- *   https://sympa.inria.fr/sympa/arc/mpfr/2022-10/msg00027.html
- *   https://gmplib.org/list-archives/gmp-bugs/2022-October/005200.html
- */
+along with the GNU MPFR Library; see the file COPYING.LESSER.
+If not, see <https://www.gnu.org/licenses/>. */
 
 /* Needed due to the tests on HAVE_STDARG and MPFR_USE_MINI_GMP */
 #ifdef HAVE_CONFIG_H
@@ -1678,6 +1663,59 @@ large_prec_for_g (void)
   mpfr_clear (x);
 }
 
+/* On 2023-12-02, this gives
+     vasprintf.c:670: MPFR assertion failed: len <= strlen (s)
+   with assertion checking (--enable-assert), or
+     [MPFR] tests_free(): bad size 1, should be 2
+   with tests_memory_disabled = 0 (default unless mini-gmp is used),
+   which means possible memory corruption with custom memory allocators
+   that do not ignore the size parameter of the "free" function, or
+     Error in check_null
+     expected r = 1, s = { 0, 0, 1 }
+     got      r = 0, s = { 0, 1, 1 }
+   with tests_memory_disabled = 1.
+   Fixed in commits
+     390e51ef8570da4e338e9806ecaf2d022210d951 (2023-12-03)
+     3babf029fe604c08ec517ca6945a5efb155f69d1 (2023-12-13)
+*/
+static void
+check_null (void)
+{
+  int i;
+
+  for (i = 0; i < 2; i++)
+    {
+      int r;
+      char s[3] = { 1, 1, 1 };
+      char t[5] = { 1, 1, 1, 1, 1 };
+
+      r = i == 0 ? mpfr_sprintf (s, "%c", 0)
+        : mpfr_snprintf (s, 2, "%c", 0) ;
+      if (r != 1 || s[0] != 0 || s[1] != 0 || s[2] != 1)
+        {
+          printf ("Error in check_null, i = %d\n", i);
+          printf ("expected r = %d, s = { %d, %d, %d }\n",
+                  1, 0, 0, 1);
+          printf ("got      r = %d, s = { %d, %d, %d }\n",
+                  r, s[0], s[1], s[2]);
+          exit (1);
+        }
+
+      r = i == 0 ? mpfr_sprintf (t, "%c%c%c", 8, 0, 9)
+        : mpfr_snprintf (t, 4, "%c%c%c", 8, 0, 9);
+      if (r != 3 ||
+          t[0] != 8 || t[1] != 0 || t[2] != 9 || t[3] != 0 || t[4] != 1)
+        {
+          printf ("Error in check_null, i = %d\n", i);
+          printf ("expected r = %d, t = { %d, %d, %d, %d, %d }\n",
+                  3, 8, 0, 9, 0, 1);
+          printf ("got      r = %d, t = { %d, %d, %d, %d, %d }\n",
+                  r, t[0], t[1], t[2], t[3], t[4]);
+          exit (1);
+        }
+    }
+}
+
 #if defined(HAVE_LOCALE_H) && defined(HAVE_SETLOCALE)
 
 /* The following tests should be equivalent to those from test_locale()
@@ -1853,6 +1891,7 @@ main (int argc, char **argv)
   mixed ();
   check_length_overflow ();
   large_prec_for_g ();
+  check_null ();
   test_locale ();
 
   if (getenv ("MPFR_CHECK_LIBC_PRINTF"))
